service/history/execution/workflow.go

```diff
@@ -232,22 +233,13 @@ func (r *workflowImpl) failDecision(
 		return nil
 	}
 
-	if _, err := r.mutableState.AddDecisionTaskFailedEvent(
-		decision.ScheduleID,
-		decision.StartedID,
-		types.DecisionTaskFailedCauseFailoverCloseDecision,
-		nil,
-		IdentityHistoryService,
-		"",
-		"",
-		"",
-		"",
-		0,
-	); err != nil {
+	if err := FailDecision(r.mutableState, decision, types.DecisionTaskFailedCauseFailoverCloseDecision); err != nil {
 		return err
 	}
-
-	return r.mutableState.FlushBufferedEvents()
+	if scheduleNewDecision {
+		return ScheduleDecision(r.mutableState)
```

Comment:
What will happen if this crashes partway through? It'll still get lost I assume?

Reply:
what do you mean by crash?

Reply:
these are in-memory operations, so it doesn't matter.